8371164: ArrayList.addAll() optimizations

jengebr opened the pull request:
# JVM Collections Optimizations: Eliminating toArray() Performance Bottlenecks

## Summary

This PR addresses performance bottlenecks in ArrayList.addAll() and Collections.SingletonSet.toArray() methods by implementing direct optimizations that bypass inefficient intermediate allocations and abstract implementations. The optimizations target high-frequency operations identified through profiling analysis, delivering 37% performance improvements for ArrayList operations and 17-43% performance improvements for SingletonSet operations under real-world conditions where multiple collection types are used.

## Problem Context

### ArrayList.addAll() Inefficiency
ArrayList.addAll() currently calls `c.toArray()` on the source collection to avoid iterator-based copying, but this creates unnecessary intermediate array allocation when the source is also an ArrayList. The method performs:

1. Call `c.toArray()` - creates intermediate array
2. Call `System.arraycopy()` to copy from intermediate array to destination
3. Discard intermediate array

When both source and destination are ArrayList instances, this can be optimized to direct array copying.

### Collections.SingletonSet toArray() Missing Implementation
Collections.SingletonSet inherits the default `AbstractCollection.toArray()` implementation, which:

1. Creates an Object[] of the expected size
2. Iterates through the collection (1 element)
3. Ensures "expected" size is the actual size
4. Returns the array

For a single-element collection, this overhead is disproportionate to the actual work needed. Additionally, this implementation is vulnerable to call site poisoning, showing 74-118% performance degradation under megamorphic conditions.

## Optimized Methods

### ArrayList
- **`addAll(Collection<? extends E> c)`**: Added fast path for ArrayList-to-ArrayList copying using direct `System.arraycopy()` from source's internal `elementData` array, eliminating intermediate `toArray()` allocation

### Collections.SingletonSet
- **`toArray()`**: Direct implementation returning `new Object[] {element}`
- **`toArray(T[] a)`**: Direct implementation with proper array sizing and null termination per Collection contract

## Performance Impact

| Class | Method | Size | Baseline | Optimized | Improvement |
|-------|--------|------|----------|-----------|-------------|
| ArrayList | addAll | 0 | 10.149 ns/op, 40 B/op | 3.929 ns/op, 24 B/op | **61% faster, 40% less allocation** |
| ArrayList | addAll | 5 | 23.804 ns/op, 104 B/op | 14.170 ns/op, 64 B/op | **40% faster, 38% less allocation** |
| ArrayList | addAll | 75 | 65.440 ns/op, 664 B/op | 42.187 ns/op, 344 B/op | **36% faster, 48% less allocation** |
| LinkedList | addAll | 0 | 6.951 ns/op, 40 B/op | 7.001 ns/op, 40 B/op | **No change (control)** |
| LinkedList | addAll | 5 | 26.019 ns/op, 104 B/op | 25.401 ns/op, 104 B/op | **No change (control)** |
| LinkedList | addAll | 75 | 199.813 ns/op, 664 B/op | 204.001 ns/op, 664 B/op | **No change (control)** |
| SingletonSet | addAll (unpoisoned) | 1 | 18.593 ns/op, 72 B/op | 21.418 ns/op, 72 B/op | **15% slower** |
| SingletonSet | addAll (poisoned) | 1 | 48.798 ns/op, 96 B/op | 25.029 ns/op, 72 B/op | **49% faster, 25% less allocation** |

## Implementation Details

### ArrayList Fast Path
Directly accesses `src.elementData` to eliminate the intermediate `toArray()` allocation.

### SingletonSet Optimizations
Both methods provide direct array creation without intermediate collections, eliminating call site poisoning vulnerability.

## FAQ

**Q: Why use exact class check instead of instanceof for ArrayList?**
A: This prevents the fast path from becoming megamorphic if ArrayList subclasses are common, ensuring the optimization remains effective.  Also, it's possible to create a subclass where a snapshot of elementData and size would disagree with the data returned, in which case the optimization would produce incorrect results.

**Q: Are these optimizations safe?**
A: Yes. ArrayList-to-ArrayList copying maintains identical behavior, and SingletonSet.toArray() implementations follow the Collection contract exactly.

**Q: What about other Collection types?**
A: Profiling data shows these are the biggest opportunities. Others exist but are lower-priority.

**Q: How do these optimizations interact with existing code?**
A: These are internal implementation optimizations that maintain full API compatibility. No external code changes are required.

## Benchmarks

JMH benchmarks are included to validate the optimizations:
- `ArrayListBulkOpsBenchmark.addAllArrayList()`: Tests ArrayList-to-ArrayList fast path
- `ArrayListBulkOpsBenchmark.addAllSingletonSet()`: Tests SingletonSet.toArray() optimization
- `ArrayListBulkOpsBenchmark.addAllLinkedList()`: Control case using existing implementation

## Testing

Comprehensive unit tests are included.  Tier-1 tests passed.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8371164](https://bugs.openjdk.org/browse/JDK-8371164): ArrayList.addAll() optimizations (**Enhancement** - P4)


### Reviewers
 * [Oli Gillespie](https://openjdk.org/census#ogillespie) (@olivergillespie - Author) Review applies to [dd713ae0](https://git.openjdk.org/jdk/pull/28116/files/dd713ae010bb16ba2c819b62db0aa945ca30ecd0)
 * [Stuart Marks](https://openjdk.org/census#smarks) (@stuart-marks - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/28116/head:pull/28116` \
`$ git checkout pull/28116`

Update a local copy of the PR: \
`$ git checkout pull/28116` \
`$ git pull https://git.openjdk.org/jdk.git pull/28116/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 28116`

View PR using the GUI difftool: \
`$ git pr show -t 28116`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/28116.diff">https://git.openjdk.org/jdk/pull/28116.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/28116#issuecomment-3482554371)
</details>

jengebr committed:
8371164 Optimizing ArrayList.addAll()

pavelrappo commented:
A drive-by observation: this PR adds benchmarks, but does not seem to add tests.

jengebr commented:
> A drive-by observation: this PR adds benchmarks, but does not seem to add tests.

Thank you - these methods are already covered by various existing tests, but for the most part that's indirect.  Are direct ones in order?

archiecobbs commented on src/java.base/share/classes/java/util/ArrayList.java:
```diff
@@ -750,6 +750,22 @@ public void clear() {
+            Object[] a = src.elementData;
+            int numNew = src.size;
+            if (numNew == 0)
+                return false;
```
Previously `addAll()` with an empty list would increment `modCount`, whereas now it won't. Might want to move this test down after `modCount++` to avoid the unnecessary behavioral change.

cl4es reviewed (changes requested):
> Why use exact class check instead of instanceof for ArrayList?

You might add that in general it might not be correct (or even safe) to do these sorts of optimizations on an arbitrary subclass of `ArrayList`. It's possible to create a subclass where a snapshot of `elementData` and `size` would disagree with the data returned by `toArray`, for example.

olivergillespie commented on src/java.base/share/classes/java/util/Collections.java:
```diff
@@ -5253,6 +5253,18 @@ public boolean removeIf(Predicate<? super E> filter) {
+            return new Object[] {element};
+        }
+        @SuppressWarnings("unchecked")
+        public <T> T[] toArray(T[] a) {
```
Should these have `@Override`?

pavelrappo replied:
All overriding methods should, yes. Also, there is slight a mismatch between the title and the contents of the PR: it's not only about `ArrayList`.

olivergillespie commented on test/micro/org/openjdk/bench/java/util/ArrayListBulkOpsBenchmark.java:
```diff
@@ -0,0 +1,179 @@
+    public ArrayList<String> addAllLinkedList75(Blackhole bh) {
+        ArrayList<String> result = new ArrayList<>(75);
+        result.addAll(linkedListSource75);
+        bh.consume(result);
```
No need for blackholes, just return the result, it has the same effect I believe.

olivergillespie commented on test/micro/org/openjdk/bench/java/util/ArrayListBulkOpsBenchmark.java:
```diff
@@ -0,0 +1,179 @@
+@Warmup(iterations = 3, time = 1, timeUnit = TimeUnit.SECONDS)
+@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
+@Fork(value = 1, jvmArgs = { "-XX:+UseParallelGC", "-Xmx3g" })
+public class ArrayListBulkOpsBenchmark {
```
You might find a layout like this a bit neater:

```
@State(Scope.Benchmark)
public class ArrayListBulkOpsBench {
    @Param({"0", "1", "5", "75"})
    int size;

    @Param({"ArrayList", "LinkedList"})
    String type;

    @Param({ "false", "true" })
    private boolean poison;

    List<String> source;

    @Setup(Level.Trial)
    public void setup() {
        switch (type) {
            case "ArrayList" -> source = new ArrayList<>(size);
            case "LinkedList" -> source = new LinkedList<>();
        }
        for (int i = 0; i < size; i++) source.add("key" + i);
        if (poison) poisonCallSites();
    }

    @Benchmark
    public ArrayList<String> addAll() {
        ArrayList<String> result = new ArrayList<>(size);
        result.addAll(source);
        return result;
    }

    static void poisonCallSites() {
        HashMap<String, String> hashMapSource = new HashMap<>();
        TreeSet<String> treeSetSource = new TreeSet<>();
        WeakHashMap<String, String> weakHashMapSource = new WeakHashMap<>();
        for (int i = 0; i < 75; i++) {
            hashMapSource.put("key" + i, "value" + i);
            treeSetSource.add("key" + i);
            weakHashMapSource.put("key" + i, "value" + i);
        }
        // Poison ArrayList.addAll() with different Collection types
        for (int i = 0; i < 40_000; i++) {
            ArrayList<Object> temp = new ArrayList<>();
            temp.addAll(hashMapSource.entrySet());
            temp.addAll(treeSetSource);
            temp.addAll(weakHashMapSource.keySet());
        }
    }

    @State(Scope.Benchmark)
    public static class SingletonSet {
        Set<String> singletonSetSource = Collections.singleton("key");

        @Param({ "false", "true" })
        private boolean poison;

        @Setup(Level.Trial)
        public void setup() {
            if (poison) poisonCallSites();
        }

        @Benchmark
        public ArrayList<String> addAllSingletonSet() {
            ArrayList<String> result = new ArrayList<>(1);
            result.addAll(singletonSetSource);
            return result;
        }
    }
}
```

addAll will be run against all permutations of size/type/poison. addAllSingletonSet will just be run against false/true poison.

jengebr replied:
I find it much neater, thank you - adopted in the latest revision.

jengebr commented on src/java.base/share/classes/java/util/ArrayList.java:
```diff
@@ -750,6 +750,22 @@ public void clear() {
      * @throws NullPointerException if the specified collection is null
      */
     public boolean addAll(Collection<? extends E> c) {
+        if (c.getClass() == ArrayList.class) {
```
Interestingly, this hurt the fast-path about 5%, but the control case about 35%.  I'm going to leave it as-is on the basis of data.

jengebr replied:
The comment is a good idea, but splitting out the duplication may make maintenance even worse by separating the duplicate sections.  What do you think?

Also, I'll benchmark with the split to get the impact data.  Not sure I can meaningfully validate an impact to inlining.

jdlib replied:
> cl4es: Consider simplifying this to reduce code duplication:

or even simpler?

```
Object[] a = c.getClass() == ArrayList.class ? ((ArrayList<?>)c).elementData : a.toArray();
int newNum = a.length;
...
```


> jengebr: Interestingly, this hurt the fast-path about 5%, but the control case about 35%.

what is the control case, LinkedList.addAll() ?

jdlib replied:
> That variant is incorrect since `size <= a.length`, potentially adding a surprising tail of `null` elements to the list.

@cl4es right, my suggestion was incorrect, thanks.

jengebr replied:
Thank you both!  @cl4es splitting it out had zero performance impact (per the benchmark).  As you said, it's a bit subjective... how strongly do you feel?  Other opinions are welcome too.

jengebr committed:
Adding direct unit tests, minor revisions to optimizations

jengebr commented:
> There needs to be some proof that this functionality is covered. If you could list existing tests that exercise the ArrayList.addAll(ArrayList) and Collections.singleton().toArray() methods, we could simply add this bug number to them.

I added two new unit tests that thoroughly cover this.  Full set of tier-1 tests passed.

jengebr committed:
Simplifying ArrayList.addAll() fastpath

jengebr commented:
Further investigation into the duplication/de-duplication of the fastpath shows that I *cannot reproduce* the negative results - they are clearly logged, I simply can't generate new runs with the same data.  No idea why.

I've updated the code with the simplified branching, all comments are addressed.

stuart-marks commented:
@jengebr Thanks for rechecking the benchmark. The ArrayList change looks much more sensible now, and I'm glad it performs to your expectations!

Regarding tests... I see you added a new test `SingletonSetToArray`. I took a quick look and it seems to me that these paths are already tested by the Collection [MOAT](https://github.com/openjdk/jdk/blob/jdk-26%2B23/test/jdk/java/util/Collection/MOAT.java) test. This test is somewhat difficult to follow, but it tries to test all the proper assertions over all the different collection implementations. For this case it creates a singleton set at [line 201](https://github.com/openjdk/jdk/blob/jdk-26%2B23/test/jdk/java/util/Collection/MOAT.java#L201), which calls [testCollection()](https://github.com/openjdk/jdk/blob/jdk-26%2B23/test/jdk/java/util/Collection/MOAT.java#L1243), which calls [testCollection1()](https://github.com/openjdk/jdk/blob/jdk-26%2B23/test/jdk/java/util/Collection/MOAT.java#L1248), which eventually calls [checkFunctionalInvariants()](https://github.com/openjdk/jdk/blob/jdk-26%2B23/test/jdk/java/util/Collection/MOAT.java#L793). And finally at [line 815](https://github.com/openjdk/jdk/blob/jdk-26%2B23/test/jdk/java/util/Collection/MOAT.java#L815) it checks toArray() with arrays of various lengths. So it may be that `SingletonSetToArray` isn't necessary. I see it does a check with a singleton set containing null, but I'm not sure that's actually necessary.

In the changes to the `ArrayList/AddAll.java` there is a test method `testSingletonSetToArray()`. Is this necessary, especially in light of the above?

Also in `ArrayList/AddAll.java` there are test methods `testArrayListFastPath()` and `testArrayListSubclassUsesSlowPath()`. These don't actually test whether the fast path or the slow path is used (and indeed, doing so would be hard, but we might have a discussion about how that could be tested, or even whether it should be tested). Instead, they mainly test that the *results* are as expected when calls are made that go through the different code paths. Hm, it doesn't seem like MOAT tests addAll(); that seems like an oversight. Well, it seems like some simplifications that can be done. There are a few combinations of inputs:

 * source is ArrayList or ArrayList subclass
 * source is empty or non-empty
 * destination is empty or non-empty
 * source contains nulls or not

(I'm not convinced we need separate cases for null in the contents.) In every case though there should be a unified set of assertion checks over the actual result.

It might be easier to do write this as a JUnit test instead of as open code in an old-school style `main` test. Or, it could be done as some enhancements to the MOAT tests. This poses its own challenges, but fundamentally it isn't difficult; you just have to wallow in that style for a while. :-)

I didn't look too closely at the `testModCountIncrement()` test method, but I'll observe that both JUnit and MOAT have an "expect this statement to throw an exception" construct.

jengebr committed:
Cleaning up tests per PR feedback

jengebr commented:
@stuart-marks thank you for the detailed feedback!  I've updated as follows:
1. Reverted ArrayList/AddAll, deleted the singleton test.
2. Added `testAddAll(Collection<Integer> c)` to MOAT
3. I implemented mostly as you described:
  1. I agree with "I'm not convinced we need separate cases for null in the contents." and did not implement them.
  1. I verified functionality of fast- and slow-paths, but did not attempt to verify which path executed.
  2. I stuck with MOAT-style.  JUnit has plenty of advantages, but this fits as a small change to the larger test.

jengebr committed:
Whitespace fixes

stuart-marks commented on test/jdk/java/util/Collection/MOAT.java:
```diff
@@ -887,6 +887,41 @@ private static void testStringElement(Collection<Integer> c) {
 
+    private static void testAddAll(Collection<Integer> c) {
+        if (!supportsAdd(c))
+            return;
```
There's already a check for `supportsAdd()` in the testCollection1() method, which is the only caller, so it doesn't seem necessary to have one here.

stuart-marks commented on test/jdk/java/util/Collection/MOAT.java:
```diff
@@ -887,6 +887,41 @@ private static void testStringElement(Collection<Integer> c) {
+
+        // Test empty ArrayList source
+        ArrayList<Integer> emptySource = new ArrayList<>();
+        check(!c.addAll(emptySource));
```
It seems like there's an assertion missing here. We check the return value from addAll(), but not the state of the collection after the operation. Maybe something like `check(c.isEmpty())` ?

jengebr committed:
Unit test revisions

jengebr commented on test/jdk/java/util/Collection/MOAT.java:
```diff
@@ -887,6 +887,41 @@ private static void testStringElement(Collection<Integer> c) {
+        arraySource.add(42);
+        arraySource.add(99);
+        check(c.addAll(arraySource));
+        equal(new ArrayList<Integer>(c), arraySource);
```
I'm curious, why not .equals() when we know the input is a List?  That is a stricter assertion because it ensures ordering remains unchanged.

Happy to make the change, though.

stuart-marks replied:
The testCollection1() method gets called with a bunch of different collections, including a HashSet, which arrives at this code in the `c` parameter. When `c` is copied into a new ArrayList, it gets whatever order is presented by `c`, which is unspecified when `c` is a HashSet. The resulting order might differ from the order in the source list, and equals() over Lists compares order, so it might result in a spurious failure.

I note that creating a HashSet of capacity 8 and adding 42 and then 99 results in [42, 99] but doing the same with a HashSet of capacity 16 results in [99, 42]. The test might actually pass, but if it does, it seems like it's pretty brittle.

stuart-marks commented:
Test update to MOAT looks good! Please add 8371164 to the `@bugs` section at the top of the file.

jengebr committed:
Polishing test and benchmark

jengebr committed:
Removing explicit classname

stuart-marks commented:
I'm running this through our internal CI system again and I can sponsor this when the run finishes.

jengebr commented:
@stuart-marks thank you very much!  :)

olivergillespie commented:
/backport jdk25u-dev
